routes/s3handler.js

```diff
@@ -0,0 +1,239 @@
+        hashedCanonicalRequest = CryptoJS.SHA256(matches[3]),
+        stringToSign = headersStr.replace(/(.+s3\/aws4_request\n)[\s\S]+/, '$1' + hashedCanonicalRequest);
+
+    return getV4SignatureKey(clientSecretKey, matches[1], matches[2], "s3", stringToSign);
+}
+
+// Signs "simple" (non-chunked) upload requests.
+function signPolicy(req, res) {
+    var policy = req.body,
+        base64Policy = new Buffer(JSON.stringify(policy)).toString("base64"),
+        signature = req.query.v4 ? signV4Policy(policy, base64Policy) : signV2Policy(base64Policy);
+
+    var jsonResponse = {
+        policy: base64Policy,
+        signature: signature
+    };
+
+    res.setHeader("Content-Type", "application/json");
+
+    if (isPolicyValid(req.body)) {
+        res.end(JSON.stringify(jsonResponse));
+    }
+    else {
+        res.status(400);
+        res.end(JSON.stringify({invalid: true}));
+    }
+}
+
+function signV2Policy(base64Policy) {
+    return getV2SignatureKey(clientSecretKey, base64Policy);
+}
+
+function signV4Policy(policy, base64Policy) {
+    var conditions = policy.conditions,
+        credentialCondition;
+
+    for (var i = 0; i < conditions.length; i++) {
+        credentialCondition = conditions[i]["x-amz-credential"];
+        if (credentialCondition != null) {
+            break;
+        }
+    }
+
+    var matches = /.+\/(.+)\/(.+)\/s3\/aws4_request/.exec(credentialCondition);
+    return getV4SignatureKey(clientSecretKey, matches[1], matches[2], "s3", base64Policy);
+}
+
+// Ensures the REST request is targeting the correct bucket for chunking support.
+function isValidRestRequest(headerStr, version) {
+    if (version === 4) {
+        return new RegExp("host:" + expectedHostname).exec(headerStr) != null;
+    }
+
+    return new RegExp("\/" + expectedBucket + "\/.+$").exec(headerStr) != null;
+}
+
+// Ensures the policy document associated with a "simple" (non-chunked) request is
+// targeting the correct bucket and the min/max-size is as expected.
+// Comment out the expectedMaxSize and expectedMinSize variables near
+// the top of this file to disable size validation on the policy document.
+function isPolicyValid(policy) {
+    var bucket, parsedMaxSize, parsedMinSize, isValid;
+
+    policy.conditions.forEach(function(condition) {
+        if (condition.bucket) {
+            bucket = condition.bucket;
+        }
+        else if (condition instanceof Array && condition[0] === "content-length-range") {
+            parsedMinSize = condition[1];
+            parsedMaxSize = condition[2];
+        }
+    });
+
+    isValid = bucket === expectedBucket;
+
+    // If expectedMinSize and expectedMax size are not null, then
+    // ensure that the client and server have agreed upon the exact same
+    // values.
+    if (expectedMinSize != null && expectedMaxSize != null) {
+        isValid = isValid && (parsedMinSize === expectedMinSize.toString())
+            && (parsedMaxSize === expectedMaxSize.toString());
```

Comment:
Maybe break out these comparisons from line 125-132 into their own `const`s so you don't need to add the comments and it's more readable.